### PR TITLE
[FLINK-32387][runtime] InputGateDeploymentDescriptor uses cache to avoid deserializing shuffle descriptors multiple times

### PR DESCRIPTION
**Pull request by huwh**
## What is the purpose of the change
InputGateDeploymentDescriptor uses cache to avoid deserializing shuffle descriptors multiple times
The cache only affects when the shuffle descriptors are offloaded by the blob server. This means the shuffle descriptors size is large enough to use caches.

## Brief change log
  - InputGateDeploymentDescriptor tries to get the shuffle descriptors for the offloaded one.

## Verifying this change

This change is already covered by existing tests.

## Does this pull request potentially affect one of the following parts:

  - Dependencies (does it add or upgrade a dependency): (no)
  - The public API, i.e., is any changed class annotated with `@Public(Evolving)`: (no)
  - The serializers: (no)
  - The runtime per-record code paths (performance sensitive): (no)
  - Anything that affects deployment or recovery: JobManager (and its components), Checkpointing, Kubernetes/Yarn, ZooKeeper: (task deployment)
  - The S3 file system connector: (no)

## Documentation

  - Does this pull request introduce a new feature? (yes)
  - If yes, how is the feature documented? (not documented)

### CODE REVIEW

File: flink-runtime/src/main/java/org/apache/flink/runtime/deployment/SerializedShuffleDescriptorAndIndices.java [hunk 1]
```diff
@@ -0,0 +1,48 @@
...
+
+import org.apache.flink.runtime.deployment.TaskDeploymentDescriptor.MaybeOffloaded;
+import org.apache.flink.runtime.deployment.TaskDeploymentDescriptorFactory.ShuffleDescriptorAndIndex;
+
+/** Wrapper of serializedShuffleDescriptors, used for {@link InputGateDeploymentDescriptor}. */
+public class SerializedShuffleDescriptorAndIndices {
+    /** Serialized value of shuffle descriptors with index. */
+    private final MaybeOffloaded<ShuffleDescriptorAndIndex[]> serializedShuffleDescriptors;
+
+    private final SerializedShuffleDescriptorAndIndicesID
```
Comment (wanglijie95) [inline flink-runtime/src/main/java/org/apache/flink/runtime/deployment/SerializedShuffleDescriptorAndIndices.java:29]: As mentioned above, rename this to `ShuffleDescriptorGroupID `. This id is used to identify both the serialized and de-serialized shuffle descriptors.

File: flink-runtime/src/main/java/org/apache/flink/runtime/taskexecutor/ShuffleDescriptorsCache.java [hunk 2]
```diff
@@ -0,0 +1,101 @@
...
+    /** Stop cache manager. */
+    void stop();
+
+    /**
+     * Get shuffle descriptors entry in cache.
+     *
+     * @param serializedShuffleDescriptorsId of cached serialized shuffle descriptors
+     * @return cache shuffle descriptors entry in cache if exists, otherwise null
+     */
+    ShuffleDescriptorCacheEntry get(
```
Comment (wanglijie95) [inline flink-runtime/src/main/java/org/apache/flink/runtime/taskexecutor/ShuffleDescriptorsCache.java:44]: I think we return the deserialized shuffle descriptors(`ShuffleDescriptorGroup`) is enough here. We can move the `ShuffleDescriptorCacheEntry` into `DefaultShuffleDescriptorsCache` as a inline class.

**Review by wanglijie95 — COMMENTED** [review]
Thanks creating this PR @huwh. Since FLINK-32385 FLINK-32386  and FLINK-32387 are closely related, let us review all changes in this PR.

I haven't fully reviewed it yet, but I have some ideas about the `SerializedShuffleDescriptorAndIndices`(I suggest change it to `SerializedShuffleDescriptorGroup`) , I hope you can take a look first.

**Comment by huwh** [comment]
@wanglijie95, Thanks for the review. I made the changes as you suggested. PTAL

File: flink-runtime/src/main/java/org/apache/flink/runtime/taskexecutor/DefaultShuffleDescriptorsCache.java [hunk 3]
```diff
@@ -0,0 +1,206 @@
...
+import java.util.HashMap;
+import java.util.HashSet;
+import java.util.Iterator;
+import java.util.LinkedHashMap;
+import java.util.Map;
+import java.util.Set;
+import java.util.concurrent.TimeUnit;
+
+/** Default implement of {@link ShuffleDescriptorsCache}. Entries will be expired after timeout. */
+public class DefaultShuffleDescriptorsCache implements ShuffleDescriptorsCache {
```
Comment (wanglijie95) [inline flink-runtime/src/main/java/org/apache/flink/runtime/taskexecutor/DefaultShuffleDescriptorsCache.java:39]: Can we directly use the `org.apache.flink.shaded.guava31.com.google.common.cache.Cache` as the `LRUCache` ? In this way, we can greatly simplify the logic in this class.

We can hold a `Map<JobID, List<PermanentBlobKey>>` to invalidate related cache entry when job terminated.

File: flink-runtime/src/main/java/org/apache/flink/runtime/taskexecutor/ShuffleDescriptorsCache.java [hunk 4]
```diff
@@ -0,0 +1,65 @@
...
+            PermanentBlobKey blobKey,
+            TaskDeploymentDescriptorFactory.ShuffleDescriptorAndIndex[]
+                    shuffleDescriptorAndIndices);
+
+    /**
+     * Clear all cache of the Job.
+     *
+     * @param jobId of job
+     */
+    void clearCacheOfJob(JobID jobId);
```
Comment (wanglijie95) [inline flink-runtime/src/main/java/org/apache/flink/runtime/taskexecutor/ShuffleDescriptorsCache.java:64]: Maybe `clearCacheForJob`

File: flink-runtime/src/main/java/org/apache/flink/runtime/taskexecutor/TaskManagerServices.java [hunk 5]
```diff
@@ -68,6 +69,9 @@
 public class TaskManagerServices {
     private static final Logger LOG = LoggerFactory.getLogger(TaskManagerServices.class);
 
+    private static final Duration SHUFFLE_DESCRIPTORS_CACHE_TIMEOUT = Duration.ofSeconds(300);
+    private static final int SHUFFLE_DESCRIPTORS_CACHE_SIZE_LIMIT = 100;
```
[the remark this answers is not in the document]
Comment (huwh) [reply flink-runtime/src/main/java/org/apache/flink/runtime/taskexecutor/TaskManagerServices.java]: At first I chose the number of ShuffleDescriptors, but the meaning would be more complicated for the user to understand. Now I've changed it to the number of Groups

**Comment by huwh** [comment]
@wanglijie95 thanks for your review. This PR has been updated as you suggested, PTAL.

File: flink-runtime/src/main/java/org/apache/flink/runtime/deployment/InputGateDeploymentDescriptor.java [hunk 6]
```diff
@@ -139,58 +142,71 @@ public IndexRange getConsumedSubpartitionIndexRange() {
...
 
-    public void loadBigData(@Nullable PermanentBlobService blobService, JobID jobId)
+    public ShuffleDescriptor[] getShuffleDescriptors() {
+        if (inputChannels == null) {
+            throw new IllegalStateException("InputChannel should not be null.");
+        }
+        return inputChannels;
+    }
+
+    public void loadBigDataAndDeserializeShuffleDescriptors(
```
Comment (wanglijie95) [inline flink-runtime/src/main/java/org/apache/flink/runtime/deployment/InputGateDeploymentDescriptor.java:152]: I suggest changing the name to `tryLoadAndDeserializeShuffleDescriptors`

File: flink-runtime/src/main/java/org/apache/flink/runtime/taskexecutor/TaskManagerServices.java [hunk 7]
```diff
@@ -68,6 +72,28 @@
...
+     * This is an expert option, that we do not want to expose in the documentation. The default
+     * value is good enough for almost all cases
+     */
+    @Experimental
+    private static final ConfigOption<Integer> SHUFFLE_DESCRIPTORS_CACHE_SIZE_LIMIT =
+            ConfigOptions.key("taskmanager.shuffle-descriptors-cache.size-limit")
+                    .intType()
+                    .defaultValue(10)
+                    .withDescription("The size limit of shuffle descriptor caches.");
+
```
Comment (wanglijie95) [inline flink-runtime/src/main/java/org/apache/flink/runtime/taskexecutor/TaskManagerServices.java:96]: I tend not to introduce configuration options for these two parameters. We can provide a new ctor `DefaultShuffleDescriptorsCache()`, and using the default value "300" and "10" in that ctor.

**Review by wanglijie95 — COMMENTED** [review]
Thanks for update @huwh . I left some comments.

File: flink-runtime/src/main/java/org/apache/flink/runtime/deployment/TaskDeploymentDescriptor.java [hunk 8]
```diff
@@ -234,10 +235,14 @@ public AllocationID getAllocationId() {
      *
      * @param blobService the blob store to use (may be <tt>null</tt> if {@link
      *     #serializedJobInformation} and {@link #serializedTaskInformation} are non-<tt>null</tt>)
+     * @param shuffleDescriptorsCache cache of shuffle descriptors to reduce the cost of
+     *     deserialization
      * @throws IOException during errors retrieving or reading the BLOBs
      * @throws ClassNotFoundException Class of a serialized object cannot be found.
      */
-    public void loadBigData(@Nullable PermanentBlobService blobService)
+    public void loadBigDataAndDeserializeShuffleDescriptors(
```
Comment (wanglijie95) [inline flink-runtime/src/main/java/org/apache/flink/runtime/deployment/TaskDeploymentDescriptor.java:243]: I personlly think that we keep the original name `loadBigData` is ok.

**Comment by huwh** [comment]
@wanglijie95 I squash the commits and rebase to master branch. PTAL

**Review by wanglijie95 — APPROVED** [review]
Thanks @huwh ! LGTM.

**Comment by huwh** [comment]
@flinkbot run azure

**Comment by wanglijie95** [comment]
After patch the fix of [pr23042](https://github.com/apache/flink/pull/23042), all tests passed. Merged.

**Pull request closed without merging**